TestSourcesTest.test_itemStream_in_expected_range failure

Failed test: http://jenkins.hazelcast.com/job/jet-oss-pr-builder/3181/testReport/junit/com.hazelcast.jet.pipeline.test/TestSourcesTest/test_itemStream_in_expected_range_0__mode_member_/

Stacktrace:
```
java.lang.AssertionError: 

Expected: exception with message a string containing "com.hazelcast.jet.pipeline.test.AssertionCompletedException"
     but: message was "Exception in ProcessorTasklet{assertCollectedEventually#1}: java.lang.AssertionError: java.lang.AssertionError: emitted items is more than twice lower then expected itemsPerSecond"
Stacktrace was: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{assertCollectedEventually#1}: java.lang.AssertionError: java.lang.AssertionError: emitted items is more than twice lower then expected itemsPerSecond
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:373)
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.AssertionError: java.lang.AssertionError: emitted items is more than twice lower then expected itemsPerSecond
	at com.hazelcast.jet.pipeline.test.AssertionSinks$CollectingSinkWithTimer.timer(AssertionSinks.java:214)
	at com.hazelcast.function.ConsumerEx.accept(ConsumerEx.java:46)
	at com.hazelcast.jet.impl.pipeline.test.AssertionP.maybeFireTimer(AssertionP.java:67)
	at com.hazelcast.jet.impl.pipeline.test.AssertionP.tryProcess(AssertionP.java:60)
	at com.hazelcast.jet.impl.processor.ProcessorWrapper.tryProcess(ProcessorWrapper.java:103)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:283)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:268)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:242)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:366)
	... 3 more
Caused by: java.lang.AssertionError: emitted items is more than twice lower then expected itemsPerSecond
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.jet.pipeline.test.TestSourcesTest.lambda$test_itemStream_in_expected_range$ae5f3d6a$1(TestSourcesTest.java:137)
	at com.hazelcast.function.ConsumerEx.accept(ConsumerEx.java:46)
	at com.hazelcast.jet.pipeline.test.AssertionSinks$CollectingSinkWithTimer.receive(AssertionSinks.java:203)
	at com.hazelcast.function.BiConsumerEx.accept(BiConsumerEx.java:47)
	at com.hazelcast.jet.impl.pipeline.test.AssertionP.tryProcess(AssertionP.java:75)
	at com.hazelcast.jet.core.AbstractProcessor.tryProcess0(AbstractProcessor.java:187)
	at com.hazelcast.jet.core.AbstractProcessor.process0(AbstractProcessor.java:591)
	at com.hazelcast.jet.core.AbstractProcessor.process(AbstractProcessor.java:109)
	at com.hazelcast.jet.impl.processor.ProcessorWrapper.process(ProcessorWrapper.java:93)
	at com.hazelcast.jet.impl.pipeline.FunctionAdapter$AdaptingProcessor.process(FunctionAdapter.java:226)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.processInbox(ProcessorTasklet.java:416)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:291)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:286)
	... 6 more
```

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/test/TestSourcesTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/test/TestSourcesTest.java
@@ -51,7 +51,7 @@ public class TestSourcesTest extends PipelineTestSupport {
 
     @Test
     public void test_longStream() throws Throwable {
-        int itemsPerSecond = 50;
+        int itemsPerSecond = 20;
         int timeoutSeconds = 10;
         int numWindowResultsToWaitFor = 3;
 

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/test/TestSourcesTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/test/TestSourcesTest.java
@@ -51,7 +51,7 @@ public class TestSourcesTest extends PipelineTestSupport {
 
     @Test
     public void test_longStream() throws Throwable {
-        int itemsPerSecond = 250;
+        int itemsPerSecond = 50;
         int timeoutSeconds = 10;
         int numWindowResultsToWaitFor = 3;
 
@@ -125,7 +125,7 @@ public class TestSourcesTest extends PipelineTestSupport {
      */
     @Test
     public void test_itemStream_in_expected_range() throws Throwable {
-        int itemsPerSecond = 5327;
+        int itemsPerSecond = 109;
 
         p.readFrom(TestSources.itemStream(itemsPerSecond))
                 .withNativeTimestamps(0)

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/test/TestSourcesTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/test/TestSourcesTest.java
@@ -52,17 +52,29 @@ public class TestSourcesTest extends PipelineTestSupport {
     @Test
     public void test_longStream() throws Throwable {
         int itemsPerSecond = 20;
-        int timeoutSeconds = 10;
-        int numWindowResultsToWaitFor = 3;
 
         p.readFrom(TestSources.longStream(itemsPerSecond, 0))
          .withNativeTimestamps(0)
          .window(tumbling(SECONDS.toMillis(1)))
          .aggregate(counting())
-         .apply(assertCollectedEventually(timeoutSeconds, windowResults -> {
-             assertTrue("Didn't receive at least three results", windowResults.size() >= numWindowResultsToWaitFor);
-             assertTrue("Invalid items per second",
-                     windowResults.stream().skip(1).allMatch(wr -> wr.result() == itemsPerSecond));
+         .apply(assertCollectedEventually(60, windowResults -> {
+             //look at last 5 windows at most, always ignore first
+             int windowsToConsider = Math.min(5, Math.max(windowResults.size() - 1, 0));
+
+             //count the total no. of items emitted in those windows
+             int totalItems = windowResults.stream()
+                                           .skip(windowResults.size() - windowsToConsider)
+                                           .mapToInt(r -> r.result().intValue())
+                                           .sum();
+
+             //compute their average
+             double avgItems = (double) totalItems / windowsToConsider;
+
+             //compute how far the actual average is from the desired one
+             double deviationFromTarget = Math.abs(avgItems - itemsPerSecond);
+
+             assertTrue(String.format("Average items per second (%.2f) too far from target (%d)",
+                     avgItems, itemsPerSecond), deviationFromTarget <= 0.1d);
          }));
 
         expectedException.expectMessage(AssertionCompletedException.class.getName());

